sage.py

```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-  # in the labels, we are removing them from the label fields
-  with open(modelfile) as fh:
-    data = fh.read()
-  data = re.sub( r'\"label\" : \"([^\n|]*)\n([^\n]*)\"', r'"label" : "\1 \2"', data )
-  
-  data = data.replace('\n', '').replace(',,', ',')#.replace(', ,', ',')#.replace('\t', ' ')
-
-
-  data = re.sub(',+', ',', data)
-  machine = json.loads(data)
-
-
-  dfa = defaultdict(lambda: defaultdict(str))
-
-  for edge in machine["edges"]:
-      dfa[ edge["source"] ][ edge["name"] ] = (edge["target"], edge["appearances"])
-
-  for entry in machine["nodes"]:
-      dfa[ str(entry['id']) ]["type"] = "0"
-      dfa[str(entry['id']) ]["isred"] = int(entry['isred'])
-
-  return (dfa, machine)
-
-def traverse(dfa, sinks, sequence, statelist=False):
-  """Wrapper to traverse a given model with a string
-
-   Keyword arguments:
-   dfa -- loaded model
-   sequence -- space-separated string to accept/reject in dfa
-  """  
-  #print(dfa)
-  #in_main_model = set()
-  sev_sinks = set()
-  state = "0"
-  stlst = ["0"]
-  #print('This seq', sequence.split(" "))
-  for event in sequence.split(" "):
-      sym = event.split(":")[0]
-      sev = rev_smallmapping[sym.split('|')[0]]
-      
-      #print('curr symbol ', sym, 'state no.', dfa[state][sym]) 
-      
-      state = dfa[state][sym]
-      isred = 0
-      
-      if state != "":
-         isred = dfa[state[0]]["isred"]
-      #print(state)
-      #if state != "":
-      #if isred == 1:
-      #      in_main_model.add(state[0])
-      if state == "":
-          # return -1 for low sev symbols
-          #print(sev)
-          if len(str(sev)) >= 2:
-                #print('high-sev sink found', sev)
-                #print(stlst[-1], sinks[stlst[-1]], stlst)
+        command += ["--" + key + "=" + kwargs[key]]
+
+    result = subprocess.run(["FlexFringe/flexfringe"] + command + [args[0]], stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True)
+    print(result.returncode, result.stdout, result.stderr)
+
+
+# Step 6.1: Load the resulting S-PDFA model
+def load_model(model_file):
+    """Wrapper to load resulting model json file
+
+    Keyword arguments:
+    model_file -- path to the json model file
+    """
+
+    # Because users can provide unescaped new lines breaking json conventions
+    #   in the labels, we are removing them from the label fields
+    with open(model_file) as fh:
+        model_data = fh.read()
+        model_data = re.sub(r'\"label\" : \"([^\n|]*)\n([^\n]*)\"', r'"label" : "\1 \2"', model_data)
+
+    model_data = model_data.replace('\n', '').replace(',,', ',')
+    model_data = re.sub(',+', ',', model_data)
+    machine = json.loads(model_data)
+
+    dfa = defaultdict(lambda: defaultdict(str))
+    for edge in machine["edges"]:
+        dfa[edge["source"]][edge["name"]] = edge["target"]
+
+    # Even though the properties might not be needed, the node has to be present in the dictionary (for sinks)
+    for entry in machine["nodes"]:
```

Comment:
So, as the comment says...

Reply:
We need to check with Sicco for this one.

Reply:
I think the point here was that as soon as there is a transition from a non-sink node to a sink node, then this sink has to be defined in the `sinks` variable (in the `_traverse` method). I once removed the "nodes" part here, but got some states in AGs that used to be sinks but were not anymore. So I have kept this part here, just that I have added the properties that might be useful (at least they were useful for me during RP, even though I checked them in the JSON files from FlexFringe)